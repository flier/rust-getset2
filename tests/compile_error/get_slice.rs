use getset2::Getter;

#[derive(Getter)]
pub struct Struct {
    #[get(slice)] // #[get(slice)] is not allowed for a field that is not a Vec<T> or array [T; N]
    field: usize,
}

#[derive(Getter)]
#[get(slice)] // #[get(slice)] is ignored when it applied to the structure
pub struct Struct2 {
    field: usize,
}

fn main() {}
