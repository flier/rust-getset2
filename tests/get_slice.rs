use getset2::Getter;

#[derive(Default, Getter)]
struct Foo {
    /// `pub fn vec_field(&self) -> &[u8]`
    /// `pub fn vec_field_mut(&mut self) -> &mut [u8]`
    #[get(slice, mut_slice)]
    pub vec_field: Vec<u8>,

    /// `pub fn array_field(&self) -> &[u8]`
    /// `pub fn array_field_mut(&mut self) -> &mut [u8]`
    #[get(slice, mut_slice)]
    pub array_field: [u8; 4],
}

#[test]
fn get_slice() {
    let mut foo = Foo::default();

    foo.vec_field = vec![0; 3];
    foo.vec_field_mut().copy_from_slice(&[1, 2, 3]);

    foo.array_field_mut()
        .copy_from_slice(0x12345678u32.to_le_bytes().as_slice());

    assert_eq!(foo.vec_field(), &[1, 2, 3]);
    assert_eq!(foo.array_field(), &[120, 86, 52, 18]);
}
