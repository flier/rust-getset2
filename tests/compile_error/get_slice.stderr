error: #[get(slice)] should be applied to a Vec<T> or an array [T; N] type
 --> tests/compile_error/get_slice.rs:6:12
  |
6 |     field: usize,
  |            ^^^^^
